web_client/js/models/widget.js

```diff
@@ -0,0 +1,326 @@
+        min = min || 0;
+        mod = (value - min) / step;
+        if (step > 0 && Math.abs(Math.round(mod) - mod) > eps) {
+            return 'Value does not satisfy step "' + step + '"';
+        }
+    },
+
+    /**
+     * Validate a widget that selects a girder model.
+     * @note This method is synchronous, so it cannot validate
+     * the model on the server.
+     */
+    _validateGirderModel: function (model) {
+        var parent;
+        if (!model.value) {
+            return 'Empty value';
+        }
+
+        switch (this.get('type')) {
+            case 'new-file':
+                parent = model.parent;
+                if (!parent || parent.resourceName !== 'folder') {
+                    return 'Invalid parent model';
+                }
+                break;
+            // other model types...
+        }
+    },
+
+    /**
+     * True if the value should be coerced as a number.
+     */
+    isNumeric: function () {
+        return _.contains(
+            ['range', 'number', 'number-vector', 'number-enumeration'],
+            this.get('type')
+        );
+    },
+
+    /**
+     * True if the value should be coerced as a boolean.
+     */
+    isBoolean: function () {
+        return this.get('type') === 'boolean';
+    },
+
+    /**
+     * True if the value is a 3 component vector.
+     */
+    isVector: function () {
+        return _.contains(
+            ['number-vector', 'string-vector'],
+            this.get('type')
+        );
+    },
+
+    /**
+     * True if the value should be coerced as a color.
+     */
+    isColor: function () {
+        return this.get('type') === 'color';
+    },
+
+    /**
+     * True if the value should be chosen from one of several "values".
+     */
+    isEnumeration: function () {
+        return _.contains(
+            ['number-enumeration', 'string-enumeration'],
+            this.get('type')
+        );
+    },
+
+    /**
+     * True if the value represents a model stored in a girder
+     * collection/folder/item hierarchy.
+     */
+    isGirderModel: function () {
+        return _.contains(
+            ['directory', 'file', 'new-file', 'image'],
+            this.get('type')
+        );
+    },
+
+    /**
+     * True if the value represents a file stored in girder.
+     */
+    isFile: function () {
+        return this.get('type') === 'file';
+    },
+
+    /**
+     * Get a normalized representation of the widget's value.
+     */
+    value: function () {
+        return this.get('value');
+    },
+
+    /**
+     * A list of valid widget types.
+     */
+    types: [
+        'color',
+        'range',
+        'number',
+        'boolean',
+        'string',
+        'number-vector',
+        'string-vector',
+        'number-enumeration',
+        'string-enumeration',
+        'file',
+        'directory',
+        'new-file',
+        'image'
+    ]
+});
+
+slicer.collections.Widget = Backbone.Collection.extend({
+    model: slicer.models.Widget,
+
```

Comment:
The version in HistomicsTK used a localStorage value, but it isn't used here.  If it should be used, it needs to be added.  If it shouldn't, we should probably get rid of localStorage from the package.json file.